README.md
# Redux Clerk

[![CircleCI][build-badge]][build]
[![npm package][npm-badge]][npm]

Redux Clerk handles the async CRUD in your Redux App.

* Provides a [set of action creators](docs/ActionsCreators.md) for both asynchronous and synchronous actions.
* Provides an [extendable reducer](docs/Reducer.md).
* Handles derived datasets and provides [selectors](docs/Selectors.md) for computing derived data.
* Stores minimum possible state.
* Optionally handles optimistic updates to the store.
* State is managed and stored as an [Immutable.js](https://facebook.github.io/immutable-js/) data structure.

```
// The tidy, minimal state managed by Redux Clerk.
{

  // Full data objects are only stored once and never duplicated.
  raw: {
    '123': { uid: 123, name: 'Apple' },
    '234': { uid: 234, name: 'Banana' },
    '345': { uid: 345, name: 'Peach' }
  },

  // Redux Clerk stores derived datasets as Lists of UIDs.
  instances: {
    myTypeaheadDataset: ['234', '123'],
    myTableDataset: ['345', '234'],
    myListDataset: ['123', '234', '345']
  }
}
```

## Installation

`npm install redux-clerk --save`

## Configuration & Docs
  - [Action Creators](docs/ActionsCreators.md)
  - [Reducer](docs/Reducer.md)
  - [Selectors](docs/Selectors.md)

## Example
An example TodoMVC using Redux Clerk is available in the [example](/example) directory. To run the example:

```
git clone git@github.com:GetAmbassador/redux-clerk.git
cd redux-clerk
npm install
npm run build
cd example
npm install react-scripts -g
npm install
npm start
```

## [FAQ](docs/FAQ.md)
  - [How can I extend the reducer provided by Redux Clerk?](docs/FAQ.md#how-can-i-extend-the-reducer-provided-by-redux-clerk)
  - [How does redux-clerk know how to normalize my data?](docs/FAQ.md#how-does-redux-clerk-know-how-to-normalize-my-data)
  - [How do I maintain the sort order of the fetch response?](docs/FAQ.md#how-do-i-maintain-the-sort-order-of-the-fetch-response)
  - [How do I recompute the derived companyTypeahead dataset?](docs/FAQ.md#how-do-i-recompute-the-derived-companytypeahead-dataset)

## License
MIT

[build-badge]: https://img.shields.io/circleci/token/9b304b29e1beb2dcd1a8ca0a6a66dc4914340796/project/github/GetAmbassador/redux-clerk/master.svg?style=flat-square
[build]: https://circleci.com/gh/GetAmbassador/redux-clerk

[npm-badge]: https://img.shields.io/npm/v/redux-clerk.svg?style=flat-square
[npm]: https://www.npmjs.org/package/redux-clerk


CHANGELOG.md
## 1.1.1
###### _January 17, 2020_

- Fix deep merge with immutable lists ([#57](https://github.com/GetAmbassador/redux-clerk/pull/57))

## 1.1.0
###### _December 21, 2017_

- Pass dispatch to async handlers via thunk ([#53](https://github.com/GetAmbassador/redux-clerk/pull/53))

## 1.0.0
###### _August 10, 2017_

- Convert responseData to Immutable data structure before adding to store ([#51](https://github.com/GetAmbassador/redux-clerk/pull/51))

## 1.0.0-rc.4
###### _May 2, 2017_

- Accept responseData in handleSuccess for the update action ([#48](https://github.com/GetAmbassador/redux-clerk/pull/48))
- Update docs to reflect optional optimism ([#47](https://github.com/GetAmbassador/redux-clerk/pull/47))

## 1.0.0-rc.3
###### _April 10, 2017_

- Cast all raw ids to strings ([#44](https://github.com/GetAmbassador/redux-clerk/pull/44))
- Update selectors to handle optional optimism ([#43](https://github.com/GetAmbassador/redux-clerk/pull/43))
- Make optimistic updates optional ([#41](https://github.com/GetAmbassador/redux-clerk/pull/41))

## 1.0.0-rc.2
###### _March 23, 2017_

- Added pre action hooks ([#38](https://github.com/GetAmbassador/redux-clerk/pull/38))

## 1.0.0-rc.1
###### _March 22, 2017_

- Added post action hooks ([#36](https://github.com/GetAmbassador/redux-clerk/pull/36))
- Rename delete action to remove to avoid reserved word conflicts ([#35](https://github.com/GetAmbassador/redux-clerk/pull/35))

## 0.3.0
###### _March 15, 2017_

- Fetch response will now deep merge response data into raw data ([#33](https://github.com/GetAmbassador/redux-clerk/pull/33))

## 0.2.8
###### _February 21, 2017_

- Cast uid to int in record selector ([#31](https://github.com/GetAmbassador/redux-clerk/pull/31))

## 0.2.7
###### _February 16, 2017_

- Added the get selector ([#29](https://github.com/GetAmbassador/redux-clerk/pull/29))

## 0.2.6
###### _January 11, 2017_

- Added the datasetProperty selector ([#27](https://github.com/GetAmbassador/redux-clerk/pull/27))

## 0.2.5
###### _January 3, 2017_

- Return Map instead of List for empty dataset ([#25](https://github.com/GetAmbassador/redux-clerk/pull/25))
- Ensure data is properly converted between actions and reducers ([#24](https://github.com/GetAmbassador/redux-clerk/pull/24))

## 0.2.4
###### _December 15, 2016_

- uidField option has been removed for the reducer (it is not longer needed) ([#22](https://github.com/GetAmbassador/redux-clerk/pull/22))

## 0.2.3
###### _December 14, 2016_

- Records are now added while updating if not already in state ([#20](https://github.com/GetAmbassador/redux-clerk/pull/20))

## 0.2.2
###### _December 13, 2016_

- Added the ability to return fetch response as an object or array ([#18](https://github.com/GetAmbassador/redux-clerk/pull/18))

## 0.2.1
###### _December 7, 2016_

- Added the ability to pass additional data with fetch, create and delete actions ([#16](https://github.com/GetAmbassador/redux-clerk/pull/16))

## 0.2.0
###### _December 5, 2016_

- Initial Release


src/reducer.js
import Immutable, { Map } from 'immutable'
import { generateActionNames } from './utils/actionNames'
import { createReducer, fetchReducer, updateReducer, deleteReducer } from './reducers'

/**
 * Creates a reducer with the provided config
 * @param  {Object} config - reducer configuration
 *
 * @return {Function} - Reducer function
 */
const reducer = (config) => {

  if (!config) throw new Error('clerk.reducer: Expected config')
  if (!config.actionPrefix) throw new Error('clerk.reducer: Expected actionPrefix')
  if (!config.uidField) throw new Error('clerk.reducer: Expected uidField')

  // Generate action names for create, fetch, update, delete
  const createActions = generateActionNames(config.actionPrefix, 'create')
  const fetchActions = generateActionNames(config.actionPrefix, 'fetch')
  const updateActions = generateActionNames(config.actionPrefix, 'update')
  const deleteActions = generateActionNames(config.actionPrefix, 'delete')

  const defaultState = Immutable.fromJS({

    // raw: Contains raw data objects
    // {
    //   123: { ... },
    //   234: { ... }
    // }
    raw: {},

    // Holds derived subsets of the raw data
    // Stored as arrays of uids
    // {
    //   instance1: { data: [123, 234] },
    //   instance2: { [234,345,456] }
    // }
    instances: {},

    // pendingDelete: Contains items that are pending deletion
    pendingDelete: {},

    // pendingUpdate: Contains items that are pending update
    pendingUpdate: {}
  })

  // Clerk reducer
  return (state = new Map({}), action) => {

    // Extend existing state
    // This comes into play when clerk is used to extend an existing reducer
    state = state.merge(defaultState)

    switch(action.type) {
      case fetchActions.start: return fetchReducer.start(state, action)
      case fetchActions.success: return fetchReducer.success(state, action)
      case fetchActions.error: return fetchReducer.error(state, action)
      case createActions.start: return createReducer.start(state, action)
      case createActions.success: return createReducer.success(state, action)
      case createActions.error: return createReducer.error(state, action)
      case updateActions.start: return updateReducer.start(state, action)
      case updateActions.success: return updateReducer.success(state, action)
      case updateActions.error: return updateReducer.error(state, action)
      case deleteActions.start: return deleteReducer.start(state, action)
      case deleteActions.success: return deleteReducer.success(state, action)
      case deleteActions.error: return deleteReducer.error(state, action)
      default: return state
    }
  }
}

export default reducer
